plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c

```diff
@@ -0,0 +1,374 @@
+		stream->pipe = NULL;
+
+		os_atomic_set_bool(&stream->active, false);
+		os_atomic_set_bool(&stream->sent_headers, false);
+
+		info("Output of file '%s' stopped",
+		     stream->printable_path.array);
+	}
+
+	if (code) {
+		obs_output_signal_stop(stream->output, code);
+	} else if (stopping(stream)) {
+		obs_output_end_data_capture(stream->output);
+	}
+
+	os_atomic_set_bool(&stream->stopping, false);
+
+	if (stream->mux_thread_joinable) {
+		os_event_signal(stream->stop_event);
+		os_sem_post(stream->write_sem);
+		pthread_join(stream->mux_thread, NULL);
+		stream->mux_thread_joinable = false;
+	}
+
+	struct encoder_packet *packet;
+	size_t i;
+
+	pthread_mutex_lock(&stream->write_mutex);
+	for (i = 0; i < stream->mux_packets.num; i++) {
+		packet = &stream->mux_packets.array[i];
+		obs_encoder_packet_release(packet);
+	}
+	da_free(stream->mux_packets);
+	pthread_mutex_unlock(&stream->write_mutex);
+
+	return ret;
+}
+
+void ffmpeg_hls_mux_destroy(void *data)
+{
+	struct ffmpeg_muxer *stream = data;
+
+	if (stream) {
+		if (stream->mux_thread_joinable)
+			pthread_join(stream->mux_thread, NULL);
+
+		hls_deactivate(stream, 0);
+
+		pthread_mutex_destroy(&stream->write_mutex);
+		os_sem_destroy(stream->write_sem);
+		os_event_destroy(stream->stop_event);
+
+		os_process_pipe_destroy(stream->pipe);
+		dstr_free(&stream->path);
+		dstr_free(&stream->printable_path);
+		dstr_free(&stream->stream_key);
+		dstr_free(&stream->muxer_settings);
+		bfree(data);
+	}
+}
+
+void *ffmpeg_hls_mux_create(obs_data_t *settings, obs_output_t *output)
+{
+	struct ffmpeg_muxer *stream = bzalloc(sizeof(*stream));
+	pthread_mutex_init_value(&stream->write_mutex);
+	stream->output = output;
+
+	/* init mutex, semaphore and event */
+	if (pthread_mutex_init(&stream->write_mutex, NULL) != 0)
+		goto fail;
+	if (os_event_init(&stream->stop_event, OS_EVENT_TYPE_AUTO) != 0)
+		goto fail;
+	if (os_sem_init(&stream->write_sem, 0) != 0)
+		goto fail;
+
+	if (obs_output_get_flags(output) & OBS_OUTPUT_SERVICE)
+		stream->is_network = true;
+
+	UNUSED_PARAMETER(settings);
+	return stream;
+
+fail:
+	pthread_mutex_destroy(&stream->write_mutex);
+	os_sem_destroy(stream->write_sem);
+	os_event_destroy(stream->stop_event);
+	bfree(stream);
+	return NULL;
+}
+
+static int process_packet(struct ffmpeg_muxer *stream)
+{
+	struct encoder_packet *packet;
+	int remaining;
+
+	pthread_mutex_lock(&stream->write_mutex);
+	remaining = stream->mux_packets.num;
+	if (remaining) {
+		packet = &stream->mux_packets.array[0];
+		write_packet(stream, packet);
+		obs_encoder_packet_release(packet);
+		da_erase(stream->mux_packets, 0);
+	}
+	pthread_mutex_unlock(&stream->write_mutex);
+
+	return 0;
+}
+
+static void *write_thread(void *data)
+{
+	struct ffmpeg_muxer *stream = data;
+	stream->active = true;
+
+	while (os_sem_wait(stream->write_sem) == 0) {
+
+		if (os_event_try(stream->stop_event) == 0)
+			break;
+
+		int ret = process_packet(stream);
```

Comment:
process_packet doesn't really return these different status. Can be just a bool.